packages/agents/lighthouse/src/tasks/prover/operations/publisher.ts

```diff
@@ -22,11 +23,93 @@ import { DEFAULT_PROVER_BATCH_SIZE, DEFAULT_PROVER_PUB_MAX } from "../../../conf
 
 import { BrokerMessage, PROVER_QUEUE } from "./types";
 
+const DEFAULT_REPLICA_SYNC_TIME = 300; // 5 min
+/**
+ * Pre fetches the unprocessed messages from the database and store them into the `messages` cache
+ */
+export const prefetch = async () => {
+  const { requestContext, methodContext } = createLoggingContext(prefetch.name);
+  const {
+    logger,
+    adapters: { database, cache },
+    config,
+  } = getContext();
+
+  // Only process configured chains.
+  const domains: string[] = Object.keys(config.chains);
+  for (const originDomain of domains) {
+    const cachedNonce = await cache.messages.getNonce(originDomain);
+    logger.info("Getting unprocessed messages from database", requestContext, methodContext, {
+      originDomain,
+      startIndex: cachedNonce + 1,
+    });
+
+    const unprocessed: XMessage[] = await database.getUnProcessedMessages(originDomain, 1000, 0, cachedNonce + 1);
+    const indexes = unprocessed.map((item: XMessage) => item.origin.index);
+    if (indexes.length > 0) {
+      logger.info(
+        "Stored unprocessed messages in the cache",
+        requestContext,
+        methodContext,
+
+        {
+          originDomain,
+          startIndex: cachedNonce + 1,
+          min: Math.min(...indexes),
+          max: Math.max(...indexes),
+        },
+      );
+
+      await cache.messages.storeMessages(unprocessed);
+      await cache.messages.setNonce(originDomain, Math.max(...indexes));
+    }
+  }
+};
+
+/**
+ * Gets unprocessed messages from the cache for a given domain pair.
+ * @param originDomain - The origin domain.
+ * @param destinationDomain - The destination domain.
+ * @param endIndex - The upper bound
+ */
+export const getUnProcessedMessagesByIndex = async (
+  originDomain: string,
+  destinationDomain: string,
+  endIndex: number,
+): Promise<XMessage[]> => {
+  const {
+    config,
+    adapters: { cache },
+  } = getContext();
+  const pendingMessages: XMessage[] = [];
+  const waitTime = config.messageQueue.publisherWaitTime ?? DEFAULT_REPLICA_SYNC_TIME;
+  let offset = 0;
+  const limit = 1000;
+  let end = false;
+  while (!end) {
+    const leaves = await cache.messages.getPending(originDomain, destinationDomain, offset, limit);
+    for (const leaf of leaves) {
+      const message = await cache.messages.getMessage(leaf);
+      if (
+        message &&
+        getNtpTimeSeconds() - message.timestamp > waitTime * 2 ** message.attempt &&
```

Comment:
Consider this scenario:
There is a long backlog in the queue > `waitTime * 2 ** message.attempt`. Since only consumer will bump the attempt count, after consuming, publisher will include the message in the batch, in every iteration, and create duplicates.

Solution could be, consumer check if it is still pending, and if so does the same timestamp check as publisher, before consuming the message. So that the exponential back off is applied

Reply:
Fixed